src/psyclone/psyir/transformations/matmul2code_trans.py

```diff
@@ -0,0 +1,286 @@
+                            "onwards of the first (matrix) argument should "
+                            "not be Ranges but found {0} at index {1}."
+                            "".format(type(index).__name__, count+2))
+
+        if len(vector.symbol.shape) == 1 and not vector.children:
+            # If the vector only has 1 dimension and all of its data is
+            # used in the matrix vector multiply then the reference does
+            # not need to supply any dimension information.
+            pass
+        else:
+            # There should be one index per dimension. This is enforced
+            # by the array create method so is not tested here.
+
+            # The first index should be a range. This
+            # transformation is currently limited to Ranges which
+            # specify the full extent of the dimension.
+            if not vector.is_full_range(0):
+                raise GenerationError(
+                    "To use matmul2code_trans on matmul, index 0 of the 2nd "
+                    "(vector) argument '{0}' must be a full range."
+                    "".format(matrix.name))
+            if len(vector.children) > 1:
+                # The 2nd index and onwards must be references.
+                for (count, index) in enumerate(vector.children[1:]):
+                    if isinstance(index, Range):
+                        raise GenerationError(
+                            "To use matmul2code_trans on matmul, indices 1 "
+                            "onwards of the second (vector) argument should "
+                            "not be Ranges but found {0} at index {1}."
+                            "".format(type(index).__name__, count+1))
+
+    def apply(self, node, options=None):
+        '''Apply the MATMUL intrinsic conversion transformation to the
+        specified node. This node must be a MATMUL
+        BinaryOperation. Currently only the matrix vector version of
+        MATMUL is supported. If the transformation is successful then
+        an assignment which includes a MATMUL BinaryOperation node
+
+        .. code-block:: fortran
+
+            R = MATMUL(A,B)
+
+        is converted to the following equivalent inline code:
+
+        .. code-block:: fortran
+
+            do i=1,N
+                R(i) = 0.0
+                    do j=1,M
+                        R(i) = R(i) + A(i,j) * B(j)
+
+        :param node: a MATMUL Binary-Operation node.
+        :type node: :py:class:`psyclone.psyGen.BinaryOperation`
+        :param options: a dictionary with options for transformations.
+        :type options: dictionary of string:values or None
+
+        '''
+        self.validate(node)
+
+        assignment = node.parent
+        matrix = node.children[0]
+        vector = node.children[1]
+        result = node.parent.lhs
+        result_symbol = result.symbol
+
+        # Create new i and j loop iterators
+        symbol_table = node.find_symbol_table()
```

Comment:
Now that all Schedules have a symbol table (or is this part of #596?) this may not give you the symbol table that you want. Actually, I think this is really an issue for the backend as it should check all child symbol tables when creating declarations?

Reply:
Yes, I would say it is a backend thing. However, it does not solve the problem of there already being a symbol with the same name in a different symbol table. This is the subject of #630 and I've added a reference to it in the code.

Reply:
Great.